### ISSUE TITLE
Native Rust AD: Forward-mode PoC (DualTensor + jvp)

### ISSUE THREAD

**Issue by shinaoka** [shinaoka/ndtensors-rs#55]
## Parent Issue
Part of #52 (AD Roadmap), specifically Phase 3: Native Rust AD.

## Goal

Minimal proof-of-concept for forward-mode autodiff using dual numbers.

**Scope**: Simple `DualTensor` that can:
1. Propagate tangent vectors through forward pass
2. Support `contract` operation as first example
3. Enable Hessian-vector product when combined with #54

**Depends on**: #54 (for Hessian-vector product demonstration)

### Architecture Overview

Following the same pattern as #54 (TrackedTensor), `DualTensor` uses type erasure internally while providing ITensors-compatible defaults through C-API.

```
┌─────────────────────────────────────────────────────────┐
│ Julia / Python                                          │
│   - frule compatible API                                │
│   - Default: same device as input                       │
└─────────────────────────────────────────────────────────┘
                          ↓ C-API
┌─────────────────────────────────────────────────────────┐
│ C-API Layer                                              │
│   - ndt_contract_jvp(a, da, b, db, ...) → (c, dc)       │
└─────────────────────────────────────────────────────────┘
                          ↓
┌─────────────────────────────────────────────────────────┐
│ Rust Core (type-erased)                                 │
│   - DualTensor<T> with AnyStorage                       │
│   - Consistent with TrackedTensor design                │
└─────────────────────────────────────────────────────────┘
```

### Core Data Structure

```rust
// crates/ndtensors/src/autodiff/dual.rs

/// Dual number tensor for forward-mode AD
/// Represents: primal + ε * tangent (where ε² = 0)
/// 
/// Uses type erasure (AnyStorage) for device flexibility,
/// consistent with TrackedTensor design in #54.
pub struct DualTensor<T: Scalar> {
    pub primal: Box<dyn AnyStorage<T>>,
    pub tangent: Option<Box<dyn AnyStorage<T>>>,  // None = zero tangent
}

impl<T: Scalar> DualTensor<T> {
    /// Create dual tensor with tangent
    pub fn new(primal: Box<dyn AnyStorage<T>>, tangent: Box<dyn AnyStorage<T>>) -> Self {
        Self { primal, tangent: Some(tangent) }
    }
    
    /// Create dual tensor without tangent (constant)
    pub fn constant(primal: Box<dyn AnyStorage<T>>) -> Self {
        Self { primal, tangent: None }
    }
    
    /// Create from concrete Dense tensors (convenience)
    pub fn from_dense<D: DataBuffer<T>>(
        primal: DenseTensor<T, D>,
        tangent: Option<DenseTensor<T, D>>,
    ) -> Self {
        Self {
            primal: Box::new(primal),
            tangent: tangent.map(|t| Box::new(t) as Box<dyn AnyStorage<T>>),
        }
    }
    
    /// Extract primal value
    pub fn primal(&self) -> &dyn AnyStorage<T> {
        self.primal.as_ref()
    }
    
    /// Extract tangent (JVP result)
    pub fn tangent(&self) -> Option<&dyn AnyStorage<T>> {
        self.tangent.as_ref().map(|t| t.as_ref())
    }
}
```

### JVP Operations

```rust
/// Forward-mode contract: compute (C, dC) from (A, dA), (B, dB)
/// 
/// JVP rule (Leibniz/product rule):
///   dC = contract(dA, B) + contract(A, dB)
pub fn dual_contract<T: Scalar>(
    a: &DualTensor<T>,
    labels_a: &[i32],
    b: &DualTensor<T>,
    labels_b: &[i32],
) -> Result<DualTensor<T>, TensorError> {
    // Primal: C = contract(A, B)
    let c_primal = contract_any(a.primal(), labels_a, b.primal(), labels_b)?;
    
    // Tangent: dC = contract(dA, B) + contract(A, dB)
    let c_tangent = match (&a.tangent, &b.tangent) {
        (Some(da), Some(db)) => {
            let term1 = contract_any(da.as_ref(), labels_a, b.primal(), labels_b)?;
            let term2 = contract_any(a.primal(), labels_a, db.as_ref(), labels_b)?;
            Some(add_any(&term1, &term2)?)
        }
        (Some(da), None) => {
            Some(contract_any(da.as_ref(), labels_a, b.primal(), labels_b)?)
        }
        (None, Some(db)) => {
            Some(contract_any(a.primal(), labels_a, db.as_ref(), labels_b)?)
        }
        (None, None) => None,
    };
    
    Ok(DualTensor { primal: c_primal, tangent: c_tangent })
}
```

### C-API Design

```c
// JVP for contraction: compute both primal and tangent
// If tangent inputs are NULL, treated as zero
int32_t ndt_contract_jvp_f64(
    // Primal inputs
    const NdtTensorF64* a,
    const NdtTensorF64* b,
    const int32_t* labels_a,
    size_t labels_a_len,
    const int32_t* labels_b,
    size_t labels_b_len,
    // Tangent inputs (NULL = zero)
    const NdtTensorF64* tangent_a,  // Can be NULL
    const NdtTensorF64* tangent_b,  // Can be NULL
    // Outputs
    NdtTensorF64** primal_out,
    NdtTensorF64** tangent_out      // NULL if both tangent inputs are NULL
);
```

### Julia Usage Example

```julia
# Forward-mode AD (frule)
function ChainRulesCore.frule((_, Δa, _, Δb, _), ::typeof(ndt_contract), a, labels_a, b, labels_b)
    # Call JVP C-API
    c, Δc = ccall(:ndt_contract_jvp_f64, ...,
                  a, b, labels_a, labels_b,
                  Δa, Δb)  # tangents (can be ZeroTangent → NULL)
    return c, Δc
end
```

### Pure Rust Usage Example

```rust
use ndtensors::autodiff::{DualTensor, dual_contract};

// Primal inputs
let a = DenseTensor::from_vec(vec![1.0, 2.0, 3.0, 4.0], &[2, 2]).unwrap();
let b = DenseTensor::ones(&[2, 2]);

// Tangent vectors (direction of differentiation)
let da = DenseTensor::ones(&[2, 2]);

// Create dual tensors
let a_dual = DualTensor::from_dense(a, Some(da));
let b_dual = DualTensor::from_dense(b, None);  // db = 0

// Forward pass with JVP
let c_dual = dual_contract(&a_dual, &[1, -1], &b_dual, &[-1, 2])?;

// Extract results
let c = c_dual.primal();      // contract(a, b)
let dc = c_dual.tangent();    // JVP: ∂C/∂A · dA
```

### Hessian-Vector Product (with #54)

See #56 for full implementation. Basic idea:

```rust
/// H @ v = JVP of gradient function
/// 
/// Forward-on-reverse composition:
/// 1. g(x) = ∇f(x)         [reverse-mode, #54]
/// 2. H@v = Jg(x) @ v      [forward-mode, #55]
pub fn hvp<T: Scalar, F>(f: F, x: &dyn AnyStorage<T>, v: &dyn AnyStorage<T>) 
    -> Result<Box<dyn AnyStorage<T>>, TensorError>
where
    F: Fn(&TrackedTensor<T>) -> TrackedTensor<T>,
{
    // Implemented in #56
    todo!()
}
```

## Implementation Tasks

- [ ] **Task 1**: `DualTensor<T>` struct using `AnyStorage<T>` (from #54)
- [ ] **Task 2**: `dual_contract` using product rule
- [ ] **Task 3**: Other dual operations (add, scale, permutedims, etc.)
- [ ] **Task 4**: C-API `ndt_contract_jvp_f64`
- [ ] **Task 5**: Tests for forward-mode AD
- [ ] **Task 6**: Integration with #54 for Hessian-vector product (#56)

## File Structure

```
crates/ndtensors/src/autodiff/
├── mod.rs
├── any_storage.rs    // AnyStorage trait (shared with #54)
├── tensor.rs         // TrackedTensor (from #54)
├── dual.rs           // DualTensor
├── backward.rs       // backward() (from #54)
└── ops/
    └── contract.rs   // tracked_contract, dual_contract, ContractBackward

crates/ndtensors-capi/src/autodiff/
├── mod.rs
├── contract.rs       // ndt_contract_backward_*, ndt_contract_jvp_*
```

## JVP Derivation for Contract

For `C = contract(A, B)` (matrix multiplication as example):
```
C[i,k] = Σ_j A[i,j] * B[j,k]

dC[i,k] = Σ_j (dA[i,j] * B[j,k] + A[i,j] * dB[j,k])
        = contract(dA, B) + contract(A, dB)
```

This is the **Leibniz rule** (product rule) for differentiation. General tensor contraction follows the same pattern.

## Comparison: TrackedTensor vs DualTensor

| Aspect | TrackedTensor (#54) | DualTensor (#55) |
|--------|---------------------|------------------|
| AD mode | Reverse (VJP/backward) | Forward (JVP) |
| Storage | `Box<dyn AnyStorage<T>>` | `Box<dyn AnyStorage<T>>` |
| Graph | Builds computation graph | No graph (eager) |
| Memory | O(graph size) | O(1) per operation |
| Gradient | Computed via `backward()` | Propagated eagerly |
| Use case | Loss → params gradients | Directional derivative, HVP |

## References

- Dual numbers: https://en.wikipedia.org/wiki/Dual_number
- JAX JVP: https://jax.readthedocs.io/en/latest/jax-101/04-advanced-autodiff.html
- ChainRulesCore frule: https://juliadiff.org/ChainRulesCore.jl/stable/rule_author/example.html
